linkerd/app/gateway/src/server.rs

```diff
@@ -75,20 +76,49 @@ impl Gateway {
             .into_inner();
 
         let discover = {
-            use profiles::GetProfile;
-
-            // Apply the gateway's allowlist to the profile discovery service.
-            let allowlist = self.config.allow_discovery.clone().into();
-            let mut profiles = profiles::WithAllowlist::new(profiles, allowlist);
-            self.outbound
-                .with_stack(protocol)
-                .push_discover(svc::mk(move |GatewayAddr(addr)| {
-                    profiles.get_profile(profiles::LookupAddr(addr.into()))
-                }))
-                .into_stack()
+            use futures::future;
+
+            let allowlist = self.config.allow_discovery.clone();
+            svc::mk(move |GatewayAddr(addr)| {
+                tracing::debug!(%addr, "Discover");
+
+                if !allowlist.matches(addr.name()) {
+                    tracing::debug!(%addr, "Address not in gateway discovery allowlist");
+                    return future::Either::Left(future::err(GatewayDomainInvalid.into()));
+                }
+
+                let profile = profiles
+                    .clone()
+                    .get_profile(profiles::LookupAddr(addr.clone().into()))
+                    .instrument(tracing::debug_span!("profiles"));
+
+                // TODO(eliza): we should probably also add the allowlist to
+                // policy resolution...
```

Comment:
This comment isn't relevant here. Is it? We just applied an allowlist above.

Reply:
yup, meant to remove this